frontend/ui/widget/keyboardlayoutdialog.lua

```diff
@@ -0,0 +1,179 @@
+            is_enter_default = true,
+            callback = function()
+                local provider = self.parent.keyboard_layout_dialog.radio_button_table.checked_button.provider
+                self.parent.keyboard:setKeyboardLayout(provider)
+                UIManager:close(self.parent.keyboard_layout_dialog)
+            end,
+        },
+    })
+
+    self.radio_button_table = RadioButtonTable:new{
+        radio_buttons = radio_buttons,
+        width = self.width * 0.9,
+        focused = true,
+        scroll = false,
+        parent = self,
+        face = self.face,
+    }
+
+    -- Buttons Table
+    self.button_table = ButtonTable:new{
+        width = self.width - 2*self.button_padding,
+        button_font_face = "cfont",
+        button_font_size = 20,
+        buttons = buttons,
+        zero_sep = true,
+        show_parent = self,
+    }
+
+    self.dialog_frame = FrameContainer:new{
+        radius = Size.radius.window,
+        bordersize = Size.border.window,
+        padding = 0,
+        margin = 0,
+        background = Blitbuffer.COLOR_WHITE,
+        VerticalGroup:new{
+            align = "center",
+            self.title_widget,
+            self.title_bar,
+            VerticalSpan:new{
+                width = Size.span.vertical_large*2,
+            },
+            CenterContainer:new{
+                dimen = Geom:new{
+                    w = self.title_bar:getSize().w,
+                    h = self.radio_button_table:getSize().h,
+                },
+                self.radio_button_table,
+            },
+            CenterContainer:new{
+                dimen = Geom:new{
+                    w = self.title_bar:getSize().w,
+                    h = Size.span.vertical_large*2,
+                },
+                LineWidget:new{
+                    background = Blitbuffer.COLOR_DARK_GRAY,
+                    dimen = Geom:new{
+                        w = self.width * 0.9,
+                        h = Size.line.medium,
+                    }
+                },
+            },
+            VerticalSpan:new{
+                width = Size.span.vertical_large*2,
+            },
+            -- buttons
+            CenterContainer:new{
+                dimen = Geom:new{
+                    w = self.title_bar:getSize().w,
+                    h = self.button_table:getSize().h,
+                },
+                self.button_table,
+            }
+        }
+    }
+
+    self[1] = CenterContainer:new{
+        dimen = Geom:new{
+            w = Screen:getWidth(),
+            h = Screen:getHeight(),
+        },
+        self.dialog_frame,
+    }
+end
+
+
+function KeyboardLayoutDialog:onShow()
+    UIManager:setDirty(self, function()
+        return "ui", self.dialog_frame.dimen
+    end)
+end
+
+function KeyboardLayoutDialog:onCloseWidget()
+    UIManager:setDirty(nil, function()
+        return "partial", self[1][1].dimen
+    end)
+    return true
+end
+
+return KeyboardLayoutDialog
```

Comment:
I think this widget should close when we tap outside, for easier getting rid of it :)
(see other widgets that have a onTapClose handler, like frontend/ui/widget/hyphenationlimits.lua).

Reply:
Possibly, I'm not sure I agree with that though. I think you might say that mostly because just tapping is a rather easy way to bring up this dialog.

I think tapping should just switch to the next layout, with special dialogs relegated to hold. I'd put activating this dialog on the east side of the swipe/popup.